File: frontend/src/features/character-sheet/systems/dnd5e/damageTypes.ts
export const DND5E_DAMAGE_TYPE_GROUPS = [
  {
    label: "Físico",
    options: [
      { value: "contundente", label: "Contundente" },
      { value: "cortante", label: "Cortante" },
      { value: "perforante", label: "Perforante" },
    ],
  },
  {
    label: "Elemental / mágico",
    options: [
      { value: "acido", label: "Ácido" },
      { value: "frio", label: "Frío" },
      { value: "fuego", label: "Fuego" },
      { value: "relampago", label: "Relámpago" },
      { value: "trueno", label: "Trueno" },
      { value: "veneno", label: "Veneno" },
    ],
  },
  {
    label: "Místico / especial",
    options: [
      { value: "fuerza", label: "Fuerza" },
      { value: "necrotico", label: "Necrótico" },
      { value: "psiquico", label: "Psíquico" },
      { value: "radiante", label: "Radiante" },
    ],
  },
] as const;

export const DND5E_DAMAGE_TYPE_VALUES = [
  "contundente",
  "cortante",
  "perforante",
  "acido",
  "frio",
  "fuego",
  "relampago",
  "trueno",
  "veneno",
  "fuerza",
  "necrotico",
  "psiquico",
  "radiante",
] as const;

export type Dnd5eDamageType = (typeof DND5E_DAMAGE_TYPE_VALUES)[number];

const DAMAGE_TYPE_ALIASES: Record<string, Dnd5eDamageType> = {
  bludgeoning: "contundente",
  slashing: "cortante",
  piercing: "perforante",
  acid: "acido",
  cold: "frio",
  fire: "fuego",
  lightning: "relampago",
  thunder: "trueno",
  poison: "veneno",
  force: "fuerza",
  necrotic: "necrotico",
  psychic: "psiquico",
  radiant: "radiante",
  contundente: "contundente",
  cortante: "cortante",
  perforante: "perforante",
  acido: "acido",
  ácido: "acido",
  frio: "frio",
  frío: "frio",
  fuego: "fuego",
  relampago: "relampago",
  relámpago: "relampago",
  trueno: "trueno",
  veneno: "veneno",
  fuerza: "fuerza",
  necrotico: "necrotico",
  necrótico: "necrotico",
  psiquico: "psiquico",
  psíquico: "psiquico",
  radiante: "radiante",
};

function slugify(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[\s_-]+/g, "");
}

export function normalizeDamageType(raw: string | undefined, fallback: Dnd5eDamageType = "contundente"): Dnd5eDamageType {
  if (!raw?.trim()) return fallback;
  const key = slugify(raw);
  const mapped = DAMAGE_TYPE_ALIASES[key];
  if (mapped) return mapped;
  if ((DND5E_DAMAGE_TYPE_VALUES as readonly string[]).includes(key)) {
    return key as Dnd5eDamageType;
  }
  return fallback;
}

export function damageTypeLabel(value: string): string {
  for (const group of DND5E_DAMAGE_TYPE_GROUPS) {
    const match = group.options.find((option) => option.value === value);
    if (match) return match.label;
  }
  return value;
}

File: backend/app/rules/dnd5e/damage_types.py
from __future__ import annotations

import re
import unicodedata

DND5E_DAMAGE_TYPE_SLUGS: frozenset[str] = frozenset(
    {
        "contundente",
        "cortante",
        "perforante",
        "acido",
        "frio",
        "fuego",
        "relampago",
        "trueno",
        "veneno",
        "fuerza",
        "necrotico",
        "psiquico",
        "radiante",
    }
)

DAMAGE_TYPE_ALIASES: dict[str, str] = {
    "bludgeoning": "contundente",
    "slashing": "cortante",
    "piercing": "perforante",
    "acid": "acido",
    "cold": "frio",
    "fire": "fuego",
    "lightning": "relampago",
    "thunder": "trueno",
    "poison": "veneno",
    "force": "fuerza",
    "necrotic": "necrotico",
    "psychic": "psiquico",
    "radiant": "radiante",
    "contundente": "contundente",
    "cortante": "cortante",
    "perforante": "perforante",
    "ácido": "acido",
    "acido": "acido",
    "frío": "frio",
    "frio": "frio",
    "fuego": "fuego",
    "relámpago": "relampago",
    "relampago": "relampago",
    "trueno": "trueno",
    "veneno": "veneno",
    "fuerza": "fuerza",
    "necrótico": "necrotico",
    "necrotico": "necrotico",
    "psíquico": "psiquico",
    "psiquico": "psiquico",
    "radiante": "radiante",
}


def _slugify(value: str) -> str:
    normalized = unicodedata.normalize("NFKD", value.strip().lower())
    without_accents = "".join(ch for ch in normalized if not unicodedata.combining(ch))
    return re.sub(r"[\s_-]+", "", without_accents)


PLACEHOLDER_DAMAGE_TYPES: frozenset[str] = frozenset({"sin tipo", "untyped", "sintype", "unknown"})


def resolve_damage_type_slug(raw: str | None) -> str | None:
    """Map a damage type to a canonical slug, or None when unknown/untyped."""
    if not raw or not str(raw).strip():
        return None
    normalized = str(raw).strip().lower()
    if normalized in PLACEHOLDER_DAMAGE_TYPES:
        return None
    key = _slugify(str(raw))
    if key in DAMAGE_TYPE_ALIASES:
        return DAMAGE_TYPE_ALIASES[key]
    if key in DND5E_DAMAGE_TYPE_SLUGS:
        return key
    return None


def normalize_damage_type(raw: str | None, *, default: str = "contundente") -> str:
    slug = resolve_damage_type_slug(raw)
    if slug is not None:
        return slug
    if not raw or not str(raw).strip():
        return default
    return default
